### ISSUE TITLE
uuid1 issue on osx ?

### ISSUE THREAD

**Issue by YeFFreY** [zyro/elixir-uuid#23]
Hello,

First thanks a lot for your library.

I'm trying to use elixir-uuid for a phoenix project. In fact, I'm using the coherence authentication lib which uses the uuid lib.
On a windows machine, no problem. But on a mac I get an error : 
`Invalid argument; Expected: <<clock_seq::14>>, <<node::48>>` when trying to call the UUID.uuid1.
Is this linked to osx or something ?
Seems to be source of the problem :

```
def uuid1(_, _, _) do
    raise ArgumentError, message:
    "Invalid argument; Expected: <<clock_seq::14>>, <<node::48>>"
end

```

Sorry I'm starting with elixir/phoenix/... so I'm learning and I might be the cause of the problem :-)

**Comment by zyro** [comment]
Can you share the versions of Elixir, Erlang, and OSX you're running? Also a bit of info on the environment, particularly network interfaces might help - the output of `:inet.getifaddrs()` from an `iex` session in particular.

**Comment by YeFFreY** [comment]
Here is the details : 

`
Erlang/OTP 19 [erts-8.0.2] [source] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]
Elixir 1.3.2
`

```
iex(1)> :inet.getifaddrs()
{:ok,
 [{'lo0',
   [flags: [:up, :loopback, :running, :multicast],
    addr: {0, 0, 0, 0, 0, 0, 0, 1},
    netmask: {65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535},
    addr: {127, 0, 0, 1}, netmask: {255, 0, 0, 0},
    addr: {65152, 0, 0, 0, 0, 0, 0, 1},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0}]},
  {'gif0', [flags: [:pointtopoint, :multicast]]}, {'stf0', [flags: []]},
  {'fw0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [0, 52, 43, 144, 177, 8, 59, 0]]},
  {'en0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [248, 26, 103, 13, 164, 20],
    addr: {65152, 0, 0, 0, 64026, 26623, 65037, 42004},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0}, addr: {192, 168, 1, 10},
    netmask: {255, 255, 255, 0}, broadaddr: {192, 168, 1, 255}]},
  {'p2p0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [10, 26, 103, 13, 164, 20]]}]}
```

**Comment by YeFFreY** [comment]
I have to add that I'm working on a hackintosh. My builtin ethernet card was not detected and it has never been an issue as I'm using an external WIFI card. So I decided to enable my builtin ethernet card (multibeast,..) and now uuid1 is working correctly.
Here is the result of :inet.getifaddrs() after that modification
```
iex(2)> :inet.getifaddrs()
{:ok,
 [{'lo0',
   [flags: [:up, :loopback, :running, :multicast],
    addr: {0, 0, 0, 0, 0, 0, 0, 1},
    netmask: {65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535},
    addr: {127, 0, 0, 1}, netmask: {255, 0, 0, 0},
    addr: {65152, 0, 0, 0, 0, 0, 0, 1},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0}]},
  {'gif0', [flags: [:pointtopoint, :multicast]]}, {'stf0', [flags: []]},
  {'en0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [144, 43, 52, 87, 67, 168]]},
  {'fw0',
   [flags: [:broadcast, :multicast], hwaddr: [0, 52, 43, 144, 177, 8, 59, 0]]},
  {'en1',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [248, 26, 103, 13, 164, 20],
    addr: {65152, 0, 0, 0, 64026, 26623, 65037, 42004},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0}, addr: {192, 168, 1, 9},
    netmask: {255, 255, 255, 0}, broadaddr: {192, 168, 1, 255}]},
  {'p2p0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [10, 26, 103, 13, 164, 20]]}]}
iex(3)> 

```

Hope this can help a bit as I suppose the lib can be modified to work in any situation ?

**Comment by zyro** [comment]
Interesting. The problem is in the first case the `'fw0'` interface's `hwaddr` is used as it's the first usable one found in the list, but it's 8 bytes rather than the typical MAC address 6 bytes. The UUID v1 node name lookup doesn't currently handle this.

I'll look into how best to handle this. Thanks for the debug info, it was very helpful!

**Comment by praveenperera** [comment]
Just upgraded to MacOS Sierra and now I'm getting the same problem as well.

**Comment by wicliff** [comment]
Can confirm. Issue started after upgrading to Sierra.

**Comment by yordis** [comment]
@zyro does this still an issue?

**Comment by xtian** [comment]
I started having this same issue after upgrading to High Sierra — was working ok on Sierra.

**Comment by zyro** [comment]
@xtian What does the output of `:inet.getifaddrs()` look like on your machine?

**Comment by xtian** [comment]
@zyro 

```
Erlang/OTP 20 [erts-9.0.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Interactive Elixir (1.5.1) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> :inet.getifaddrs()
{:ok,
 [{'lo0',
   [flags: [:up, :loopback, :running, :multicast], addr: {127, 0, 0, 1},
    netmask: {255, 0, 0, 0}, addr: {0, 0, 0, 0, 0, 0, 0, 1},
    netmask: {65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535},
    addr: {65152, 0, 0, 0, 0, 0, 0, 1},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0}]},
  {'gif0', [flags: [:pointtopoint, :multicast]]}, {'stf0', [flags: []]},
  {'XHC20', [flags: []]}, {'EHC64', [flags: []]},
  {'fw0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [0, 10, 39, 2, 0, 107, 171, 67]]},
  {'en4',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [56, 201, 134, 63, 59, 102]]},
  {'en1',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [106, 0, 3, 11, 17, 176]]},
  {'en2',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [106, 0, 3, 11, 17, 177]]},
  {'en0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [24, 101, 144, 211, 105, 67],
    addr: {65152, 0, 0, 0, 1230, 24744, 47094, 15672},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0}, addr: {10, 30, 161, 43},
    netmask: {255, 255, 240, 0}, broadaddr: {10, 30, 175, 255}]},
  {'p2p0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [10, 101, 144, 211, 105, 67]]},
  {'awdl0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [166, 237, 190, 192, 80, 230],
    addr: {65152, 0, 0, 0, 42221, 48895, 65216, 20710},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0}]},
  {'bridge0',
   [flags: [:up, :broadcast, :running, :multicast],
    hwaddr: [106, 0, 3, 11, 17, 176]]},
  {'utun0',
   [flags: [:up, :pointtopoint, :running, :multicast],
    addr: {65152, 0, 0, 0, 39092, 42127, 27730, 33993},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0},
    dstaddr: {20, 18, 15, 0}]},
  {'utun1',
   [flags: [:up, :pointtopoint, :running, :multicast],
    addr: {65152, 0, 0, 0, 63144, 3373, 6892, 12512},
    netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0},
    dstaddr: {101, 97, 109, 0}]}]}
```

**Comment by zyro** [comment]
The problem is still 64-bit hardware addresses, the [UUID v1 node spec](https://tools.ietf.org/html/rfc4122#section-4.1.6) only accounts for IEEE 802 MAC address and where none is available falls back to "a randomly or pseudo-randomly generated value". I think the most correct solution (unless I've missed an overriding spec or RFC somewhere?) is to use the first 48-bit address available, or fall back to a random value if there is none even if there might be a 64-bit address available.

If no one has any other suggestions I'll put this change together and prep a new release.

**Comment by zyro** [comment]
Should be fixed by https://github.com/zyro/elixir-uuid/commit/7bb1f09cd0e9060c401f578d4faf463489cfc9a1 and released as [1.1.8 on Hex](https://hex.pm/packages/uuid/1.1.8).

**Comment by xtian** [comment]
Thanks!
